walkeeper/src/receive_wal.rs

```diff
@@ -50,18 +60,29 @@ fn request_callback(conf: SafeKeeperConf, timelineid: ZTimelineId, tenantid: ZTe
         tenantid, timelineid, host, port, timelineid, tenantid,
     );
 
+    let mut start_pos = Lsn(0);
     loop {
-        info!(
-            "requesting page server to connect to us: start {} {}",
-            ps_connstr, callme
-        );
-        match Client::connect(&ps_connstr, NoTls) {
-            Ok(mut client) => {
-                if let Err(e) = client.simple_query(&callme) {
-                    error!("Failed to send callme request to pageserver: {}", e);
+        if let Err(TryRecvError::Disconnected) = rx.try_recv() {
+            trace!("Walstream disconnected. Exiting request_callback thread");
+            break;
+        }
+
+        // Request callback only at start or
+        // if we have some new WAL to stream
+        if let Some(lsn) = timeline.wait_for_lsn(start_pos) {
```

Comment:
That's broken: even if safekeeper doesn't have new WAL since the last check, it doesn't mean pageserver is caught up.